src/core/net/dnssd_server.cpp

```diff
@@ -0,0 +1,554 @@
+
+    SuccessOrExit(error = ProcessQuery(aMessage, *responseMessage, requestHeader));
+
+exit:
+    if (error == OT_ERROR_NONE)
+    {
+        OT_ASSERT(responseMessage != nullptr);
+        error = SendMessage(*responseMessage, aMessageInfo);
+    }
+
+    FreeMessageOnError(responseMessage, error);
+}
+
+otError Server::ProcessQuery(Message &aMessage, Message &aResponse, const Header &aRequestHeader)
+{
+    Header           responseHeader;
+    uint16_t         readOffset, nameSerializeOffset;
+    Question         question;
+    uint16_t         qtype;
+    char             name[Dns::Name::kMaxLength + 1] = "";
+    otError          error                           = OT_ERROR_NONE;
+    NameCompressInfo compressInfo;
+
+    // setup initial DNS response header
+    responseHeader.Clear();
+    responseHeader.SetResponseCode(Header::kResponseSuccess);
+    responseHeader.SetType(Header::kTypeResponse);
+    responseHeader.SetMessageId(aRequestHeader.GetMessageId());
+
+    // validate the query
+    VerifyOrExit(aRequestHeader.GetQueryType() == Header::kQueryTypeStandard,
+                 responseHeader.SetResponseCode(Header::kResponseNotImplemented));
+    VerifyOrExit(!aRequestHeader.IsTruncationFlagSet(), responseHeader.SetResponseCode(Header::kResponseFormatError));
+    VerifyOrExit(aRequestHeader.GetQuestionCount() == 1, responseHeader.SetResponseCode(Header::kResponseFormatError));
+
+    // read the query name and question
+    readOffset = sizeof(Header);
+    VerifyOrExit(OT_ERROR_NONE == Dns::Name::ReadName(aMessage, readOffset, 0, name, sizeof(name)),
+                 responseHeader.SetResponseCode(Header::kResponseFormatError));
+    VerifyOrExit(OT_ERROR_NONE == aMessage.Read(readOffset, question),
+                 responseHeader.SetResponseCode(Header::kResponseFormatError));
+
+    // add the question to the response, and save the serialize offset of name
+    nameSerializeOffset = aResponse.GetLength();
+    SuccessOrExit(error = AddQuestionToResponse(name, question, aResponse, responseHeader));
+
+    // further validate the query
+    qtype = question.GetType();
+    VerifyOrExit(qtype == ResourceRecord::kTypePtr || qtype == ResourceRecord::kTypeSrv ||
+                     qtype == ResourceRecord::kTypeTxt || qtype == ResourceRecord::kTypeAaaa,
+                 responseHeader.SetResponseCode(Header::kResponseNotImplemented));
+
+    VerifyOrExit(question.GetClass() == ResourceRecord::kClassInternet ||
+                     question.GetClass() == ResourceRecord::kClassAny,
+                 responseHeader.SetResponseCode(Header::kResponseNotImplemented));
+
+    // prepare the information for name compression
+    VerifyOrExit(OT_ERROR_NONE == PrepareCompressInfo(qtype, name, nameSerializeOffset, compressInfo),
+                 responseHeader.SetResponseCode(Header::kResponseNameError));
+
+    // resolve the question
+    SuccessOrExit(error = ResolveQuestion(name, question, responseHeader, aResponse, compressInfo));
```

Comment:
For current implementation, we will not send a response when failed to resolve a question (we send the response in `Server::HandleUdpReceive` only when `error` is `OT_ERROR_NONE`). But I think we should send a response to the client in any fail cases as soon as we verified that the request message is a DNS query (we can just ignore the request if it is not a DNS query).

Reply:
Actually we only return error when the server failure happend (e..g NoBufs). Thus, I think we can handle the errors by sending a ServerFailure RCODE.